lib/schema/array.js

```diff
@@ -359,7 +359,11 @@ SchemaArray.prototype.cast = function(value, doc, init, prev, options) {
     options = options || emptyOpts;
 
     let rawValue = utils.isMongooseArray(value) ? value.__array : value;
-    value = MongooseArray(rawValue, options.path || this._arrayPath || this.path, doc, this);
+    let path = options.path || this.path;
+    if (options.arrayPathIndex != null) {
```

Comment:
is arrayPathIndex a string or a numerical value? 

If it is a number, we should actually do Number.isInteger.

Reply:
Number. I don't think it is necessary to do an extra `isInteger()` check, because `arrayPathIndex` isn't something an end user can set. Mongoose sets that in a `for` loop.